src/listview.tsx

```diff
@@ -17,72 +66,103 @@ export type ListViewProps<A, K = A> = {
-            } else if (currentValues.length === 0) {         
-                for (let i = 0; i < nextValues.length; i++) {
-                    const nextItemKey = key(nextValues[i])
-                    const newElement = renderItem(nextItemKey, nextValues, i)
-                    if (i == 0) {
-                        H.replaceNode(controller, i, newElement)        
-                    } else {
-                        H.appendNode(controller, newElement)                        
-                    }                        
+        }
+        if (result !== currentItems) {
+            key2item.forEach((info) => {
+                const i = info.index
+                if (!result[i] || result[info.index].key !== info.key) {
+                    deletedNodes.push(info.node)
+                    key2item.delete(info.key)
                 }
+            })
 
+            const oldNodes = Array.isArray(currentItems)
+                ? currentItems.map(item => item.node)
+                : [currentItems]  // <-- placeholder
+            if (nextN > 0) {
+                const newNodes = result.map(item => item.node)
+                currentItems = result
+                H.replaceAllInPlace(controller, oldNodes, newNodes, createdNodes, deletedNodes)
             } else {
-                // 1. replace at common indices
-                for (let i = 0; i < nextValues.length && i < currentValues.length; i++) {
-                    const nextItemKey = key(nextValues[i])
-                    if (nextItemKey !== key(currentValues[i])) {
-                        //console.log("Replace element for", nextValues[i])
-                        const nextElement = renderItem(nextItemKey, nextValues, i)
-                        H.replaceNode(controller, i, nextElement)                        
-                    } else {
-                        // Key match => no need to replace
-                    }
-                }
-                // 2. add/remove nodes
-                if (nextValues.length > currentValues.length) {                    
-                    for (let i = currentValues.length; i < nextValues.length; i++) {
-                        const nextItemKey = key(nextValues[i])
-                        const newElement = renderItem(nextItemKey, nextValues, i)
-                        H.appendNode(controller, newElement)
-                    }
-                } else if (nextValues.length < currentValues.length) {
-                    for (let i = currentValues.length - 1; i >= nextValues.length; i--) {
-                        H.removeNode(controller, i, controller.currentElements[i])
-                    }                    
-                }
+                const newNodes = [H.createPlaceholder()]
+                currentItems = newNodes[0]
+                H.replaceAllInPlace(controller, oldNodes, newNodes, newNodes, deletedNodes)
             }
-        } 
-        currentValues = nextValues        
+        }
```

Comment:
The whole logic up from here was inspired by [mapElemsWithId from karet.util](https://github.com/calmm-js/karet.util/blob/7dc6ff88bc136d958addd670277023f71e813fca/src/karet.util.js#L794-L831)